### ISSUE TITLE
docs: Ambiguous "datasource is used to get a reference to the repository"

### ISSUE THREAD

**Issue by mloskot** [fluxcd/terraform-provider-flux#575]
The current https://registry.terraform.io/providers/fluxcd/flux/latest/docs/guides/github says:

https://github.com/fluxcd/terraform-provider-flux/blob/fa794b3e95994460ad87d349568664f17eb9caac/docs/guides/github.md?plain=1#L62

but there is no use of the actual [github_repository](https://registry.terraform.io/providers/integrations/github/latest/docs/data-sources/repository) or any other **data source** in the guide.

The guide only refers to (one) GitHub provider **resource**:

https://github.com/fluxcd/terraform-provider-flux/blob/fa794b3e95994460ad87d349568664f17eb9caac/docs/guides/github.md?plain=1#L75-L77

IMO, the sentence referenced at the beginning of this issue above is confusing.

BTW, this issue seems to also apply to the GitLab guide

https://github.com/fluxcd/terraform-provider-flux/blob/fa794b3e95994460ad87d349568664f17eb9caac/docs/guides/gitlab.md?plain=1#L63

I guess, the **datasource** reference may come from some old version of the examples, like in here
https://github.com/fluxcd/terraform-provider-flux/blob/fa794b3e95994460ad87d349568664f17eb9caac/examples/multi-env/main.tf#L17-L19

**Comment by swade1987** [comment]
@mloskot the example is valid as the `github_repository_deploy_key` and `gitlab_deploy_key` are deploy keys that are added to the Github or Gitlab repositories. Without these flux would not be able to connect to the repositories in question within the provider configuration (see below):

```
provider "flux" {
  git = {
    url = "ssh://git@gitlab.com/${data.gitlab_project.this.path_with_namespace}.git"
    ssh = {
      username    = "git"
      private_key = tls_private_key.flux.private_key_pem
    }
  }
}
```

The `private_key` value above is the same private key which is used as the deploy key for the repository.

**Comment by swade1987** [comment]
I thought this may explain it better ...

<img width="696" alt="Screenshot 2024-03-24 at 17 12 16" src="https://github.com/fluxcd/terraform-provider-flux/assets/594451/9201f7a0-fc00-4ca1-a70d-275eddc266cf">

1. We create a private key
2. That private key is used by the deploy key
3. That deploy key is associated with a repository (either github or gitlab)
4. The flux provider git configuration uses the same private key and can therefore access the repo.

**Comment by swade1987** [comment]
Please let me know if this makes sense, if it does I will create a PR to add some comments to the terraform code.

**Comment by mloskot** [comment]
@swade1987 
> the example is valid (...)

I did not suggest the example is not valid.
I only complained about the description accompanying the example as being confusing:

> The GitHub repository is created separatly so a datasource is used
> to get a reference to the repository.

and the example does not make use of any data source, what I explained in my report above.

Although I did grasp the idea behind use of the keys and the overall workflow, thank you very much for the further explanation.
